cpp/CMakeLists.txt

```diff
@@ -395,6 +395,14 @@ set(LIBRARY_OUTPUT_DIRECTORY "${BUILD_OUTPUT_ROOT_DIRECTORY}")
 # where to put generated binaries
 set(EXECUTABLE_OUTPUT_PATH "${BUILD_OUTPUT_ROOT_DIRECTORY}")
 
+if(CMAKE_GENERATOR STREQUAL Xcode)
+  # Xcode projects support multi-configuration builds.  This forces a single output directory
+  # when building with Xcode that is consistent with single-configuration Makefile driven build.
+  set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY_${CMAKE_BUILD_TYPE} "${BUILD_OUTPUT_ROOT_DIRECTORY}")
+  set(CMAKE_LIBRARY_OUTPUT_DIRECTORY_${CMAKE_BUILD_TYPE} "${BUILD_OUTPUT_ROOT_DIRECTORY}")
+  set(CMAKE_RUNTIME_OUTPUT_DIRECTORY_${CMAKE_BUILD_TYPE} "${BUILD_OUTPUT_ROOT_DIRECTORY}")
+endif()
```

Comment:
I think that we can always set these variables.

We should use `UPPERCASE_BUILD_TYPE` instead of `CMAKE_BUILD_TYPE` because users may specify `Debug` as `CMAKE_BUILD_TYPE`.

Reply:
> I think that we can always set these variables.

I'm a bit worried about settings these variables unconditionally and possibly breaking multi-configuration generators (MSVC?).

Reply:
AppVeyor build will answer this... Oh, we don't have a job that uses "Visual Studio XXX" generator.

OK. We can set them only for Xcode for now.